cherab/core/laser/models/laserspectrum_base.pyx

```diff
@@ -0,0 +1,125 @@
+
+from raysect.core.math.function cimport Function1D
+from raysect.optical cimport Point3D, Vector3D
+
+from cherab.core.utility import Notifier
+from cherab.core.laser.node import Laser
+import numpy as np
+cimport numpy as np
+
+from cherab.core.utility.constants cimport SPEED_OF_LIGHT, PLANCK_CONSTANT
+cdef double CONSTANT_HC = SPEED_OF_LIGHT * PLANCK_CONSTANT
+
+cdef class LaserSpectrum_base(Function1D):
+
+    def __init__(self, double min_wavelength, double max_wavelength, int bins):
+
+        super().__init__()
+
+        self._notifier = Notifier()
+
+        self._check_wavelength_validity(min_wavelength, max_wavelength)
+
+        self._min_wavelength = min_wavelength
+        self._max_wavelength = max_wavelength
+
+        self.bins = bins
+
+
+    @property
+    def min_wavelength(self):
+        return self._min_wavelength
+
+    @min_wavelength.setter
+    def min_wavelength(self, double value):
+
+        self._min_wavelength = value
+        self._update_cache()
+
+    @property
+    def max_wavelength(self):
+        return self._max_wavelength
+
+    @max_wavelength.setter
+    def max_wavelength(self, double value):
+
+        self._max_wavelength = value
+        self._update_cache()
+
+    @property
+    def bins(self):
+        return self._bins
+
+    @bins.setter
+    def bins(self, int value):
+        if not value > 0:
+            raise ValueError("Value has to be larger than 0")
+
+        self._bins = value
+        self._update_cache()
+
+    @property
+    def wavelengths(self):
+        return self._wavelengths
+
+    @property
+    def psd(self):
```

Comment:
PSD? not a very descriptive name.

Reply:
Power Spectral Density, I thought it was quite a regular abbreviation, I can change it to PowerSpectralDensity if it is more appropriate.

Reply:
I don't think PSD is a particularly uncommon term, and if it's in the class's docstring I'd be fine with it. Though if you do rename it then PEP8 would imply `power_spectral_density`.

Reply:
Unless it is particularly onerous I think it should be spelled out in full, we do that elsewhere. It helps people unfamiliar with the area learn by osmosis. Jack is correct about eh PEP8 name. If you have not read PEP8, please do: https://www.python.org/dev/peps/pep-0008/ as any commit violating pep8 won't be merged.